test/e2e/framework/endpoints/ports.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+This soak tests places a specified number of pods on each node and then
+repeatedly sends queries to a service running on these pods via
+a serivce
+*/
+
+package endpoints
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
+)
+
+// PortsByPodName is a map that maps pod name to container ports.
+type PortsByPodName map[string][]int
+
+// PortsByPodUID is a map that maps pod UID to container ports.
+type PortsByPodUID map[types.UID][]int
+
+// GetContainerPortsByPodUID returns a PortsByPodUID map on the given endpoints.
+func GetContainerPortsByPodUID(ep *v1.Endpoints) PortsByPodUID {
+	m := PortsByPodUID{}
+	for _, ss := range ep.Subsets {
+		for _, port := range ss.Ports {
+			for _, addr := range ss.Addresses {
+				containerPort := port.Port
+				if _, ok := m[addr.TargetRef.UID]; !ok {
+					m[addr.TargetRef.UID] = make([]int, 0)
+				}
+				m[addr.TargetRef.UID] = append(m[addr.TargetRef.UID], int(containerPort))
+			}
+		}
+	}
+	return m
+}
+
+func translatePodNameToUIDOrFail(c clientset.Interface, ns string, expectedEndpoints PortsByPodName) PortsByPodUID {
+	portsByUID := make(PortsByPodUID)
+
+	for name, portList := range expectedEndpoints {
+		pod, err := c.CoreV1().Pods(ns).Get(name, metav1.GetOptions{})
+		if err != nil {
+			framework.Failf("failed to get pod %s, that's pretty weird. validation failed: %s", name, err)
```

Comment:
Maybe for a future PR, we should try to avoid importing `framework` here. Maybe a better pattern is to return an err here and have the calling test run `framework.Failf`.

Reply:
Actually, if you don't mind I think it would be ideal to do this refactor now. It'll keep the dependency graph cleaner for later

Reply:
updated, ready to review